src/ert/scheduler/scheduler.py

```diff
@@ -181,32 +181,70 @@ async def execute(
         # We need to store the loop due to when calling
         # cancel jobs from another thread
         self._loop = asyncio.get_running_loop()
-        async with background_tasks() as cancel_when_execute_is_done:
-            cancel_when_execute_is_done(self._publisher())
-            cancel_when_execute_is_done(self._process_event_queue())
-            cancel_when_execute_is_done(self.driver.poll())
-            if min_required_realizations > 0:
-                cancel_when_execute_is_done(
+        scheduling_tasks = [
+            asyncio.create_task(self._publisher()),
+            asyncio.create_task(self._process_event_queue()),
+            asyncio.create_task(self.driver.poll()),
+        ]
+
+        if min_required_realizations > 0:
+            scheduling_tasks.append(
+                asyncio.create_task(
                     self._stop_long_running_jobs(min_required_realizations)
                 )
-                cancel_when_execute_is_done(self._update_avg_job_runtime())
-
-            start = asyncio.Event()
-            sem = asyncio.BoundedSemaphore(self._max_running or len(self._jobs))
-            for iens, job in self._jobs.items():
-                self._tasks[iens] = asyncio.create_task(
-                    job(start, sem, self._max_submit)
-                )
-
-            start.set()
-            results = await asyncio.gather(
-                *self._tasks.values(), return_exceptions=True
             )
-            for result in results:
-                if isinstance(result, Exception):
-                    logger.error(result)
-
-            await self.driver.finish()
+            scheduling_tasks.append(asyncio.create_task(self._update_avg_job_runtime()))
+
+        start = asyncio.Event()
+        sem = asyncio.BoundedSemaphore(self._max_running or len(self._jobs))
+        for iens, job in self._jobs.items():
+            self._tasks[iens] = asyncio.create_task(job(start, sem, self._max_submit))
+        start.set()
+
+        async def gather_realization_jobs() -> list[BaseException | None]:
+            """This handles realization jobs. It makes sure that all the tasks
```

Comment:
"This handles realization jobs" is not needed as it now is in the function name.

Reply:
yes, recommended by Prof. @pinkwah ! :smile:

Reply:
- Builtin `list[T]` type is new in Python 3.9, with `typing.List` being deprecated
- Union syntax `X | Y` (equivalent to `Union[X, Y]`) is new in Python 3.10

Because we use `from __future__ import annotations`, these expressions are never computed during runtime, so this works in Python 3.8. Mypy evaluates the types separately, and uses the latest Python standard regardless of which CPython version you're running `mypy` from.

The only place this cannot be used (until we deprecate Python 3.9) is where runtime information is important, such as dataclasses:
```py
@dataclass
class Foo:
    mylist: List[Union[int, str]]  # _must_ use Python3.8-compatible syntax

    def do_something(self) -> list[int | str]:  # Not interpreted by Python, so we can use Python3.12 syntax
        ...
```

It might cause confusion but my current take is that 80% of our code is _just_ type hinting, and the new syntax is easier to understand at a glance, so it's worth.